### PR TITLE
[Task] OS-11: Create Layout Component

### PR DESCRIPTION
**Pull request by lynnmugambi**
**What does this PR do?**
This PR adds the layout folder and components to the project.

**Description of Tasks to be completed**
Create Default Layout

**How should this be manually tested?**
First, pull then checkout to this branch

`git pull`

Switch to the OS-11-Layout-Component:

`git checkout OS-11-Layout-Component`

To add and commit new changes, run:

`gc`

**Any background context you want to provide?**
The Layout component checks if the layout has already been applied, if so, it does not re-render the component. The LayoutDefault component creates the background for the Order Status Page, can be reusable.

**What are the relevant Jira story IDs?**
[#OS-11](https://gopuff.atlassian.net/jira/software/projects/OS/boards/33/backlog?selectedIssue=OS-11)

**Screenshots (if appropriate)**
None

**Questions:**
None

### CODE REVIEW

File: src/layouts/LayoutDefault.vue [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<template>
+  <div class="layoutDefault">
```
Comment (rnkimani) [inline src/layouts/LayoutDefault.vue]: Let's name our css with hyphens rather than camel case. `layout-default`

File: src/App.vue [hunk 2]
```diff
@@ -1,9 +1,22 @@
...
-    <router-view />
+    <component :is="layout">
+      <router-view :layout.sync="layout" />
+    </component>
   </div>
 </template>
 
+<script>
+export default {
+  name: `App`,
```
Comment (rnkimani) [inline src/App.vue]: A normal string could work here. "App"

File: src/App.vue [hunk 3]
```diff
@@ -1,9 +1,22 @@
...
+  data() {
+    return {
+      layout: `div`
```
Comment (rnkimani) [inline src/App.vue]: https://github.com/gopuff/order-status/pull/3/files#r310083210

File: src/layouts/_layout.vue [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    name: {
+      type: String,
+      required: true
+    }
+  },
+  created() {
+    // Check if the layout component
+    // has already been registered.
+    if (!Vue.options.components[this.name]) {
+      Vue.component(this.name, () => import(`../layouts/${this.name}.vue`));
```
Comment (rnkimani) [inline src/layouts/_layout.vue]: Nice 👏

File: src/layouts/LayoutDefault.vue [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    <main class="layoutDefault__main">
```
Comment (rnkimani) [inline src/layouts/LayoutDefault.vue]: I think using BEM is a good idea

File: src/App.vue [hunk 6]
```diff
@@ -1,9 +1,22 @@
 <template>
   <div class="app" id="app">
-    <router-view />
+    <component :is="layout">
+      <router-view :layout.sync="layout" />
```
Comment (rnkimani) [inline src/App.vue]: I don't think I have used sync before. Nice to see a way I can implement it

File: src/layouts/LayoutDefault.vue [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+      <slot />
+    </main>
+  </div>
+</template>
+
+<style lang="scss" scoped>
+.layoutDefault {
+  display: flex;
+  width: 100vw;
```
Comment (rnkimani) [inline src/layouts/LayoutDefault.vue]: Is this being used? 
I'm asking this coz the max-width has already been set

**Comment by rnkimani** [comment]
You can also remove the margin that comes by default with the body in the main entry css file

File: src/layouts/LayoutDefault.vue (hunk 7 above)
Comment (lynnmugambi) [reply src/layouts/LayoutDefault.vue]: So I copied the styling directly from the page as given. I spotted it but realized it works for smaller screens where it fills the view width until 600px screens (which is rare)

File: src/layouts/LayoutDefault.vue (hunk 5 above)
Comment (lynnmugambi) [reply src/layouts/LayoutDefault.vue]: Yeah it really helps with readibility and quick understanding

File: src/layouts/_layout.vue (hunk 4 above)
Comment (lynnmugambi) [reply src/layouts/_layout.vue]: Thanks!

File: src/App.vue (hunk 2 above)
Comment (lynnmugambi) [reply src/App.vue]: True, I think it just depends on preference.

**Pull request merged**